### ISSUE TITLE
Multiple errors in daemon logs while executing create/start/stop commands

### ISSUE THREAD

**Issue by xyeLz** [Pterodactyl/Panel#457]
Previous behavior: My daemon used to randomly throw errors in the log for what appeared to be no reason. Services started/stopped successfully (albeit this was also inconsistent) but logs would just occur every 15-45 seconds seemingly regardless of whatever was occurring. I upgraded my daemon to the 0.4.1 develop version which resolved the _random_ errors (random being the keyword).

Behavior at 2017/05/25 at 9:23 AM EST: The daemon appears to only throw errors when it actually experiences trouble (which is great since it's not clogging up the logs and will help to identify tasks that I create that are experiencing trouble). However, functions that were previously working are now broken, including servers that were working perfectly.

Here is the behavior when trying to start an _already created_ Call of Duty 4 server. This server stopped and started perfectly yesterday. It's worth nothing that this server _does_ actually start from a server standpoint and is accessible within the game. It even outputs the necessary "done" command in the panel: `Sending heartbeat to cod4master.activision.com`. I confirmed this exists in the main.json for the service. It's just never marked as "Starting".

`{"name":"wings","hostname":"9320dd4fbd6e","pid":17,"server":"67d2f846-b348-4ece-bb6c-2b0e7ff2d65c","level":30,"msg":"Server status has been changed to STARTING","time":"2017-05-25T13:34:39.488Z","v":0}
{"name":"wings","hostname":"9320dd4fbd6e","pid":17,"server":"67d2f846-b348-4ece-bb6c-2b0e7ff2d65c","level":20,"msg":"Checking size of server folder before booting.","time":"2017-05-25T13:34:39.488Z","v":0}
{"name":"wings","hostname":"9320dd4fbd6e","pid":17,"server":"67d2f846-b348-4ece-bb6c-2b0e7ff2d65c","level":20,"msg":"Initializing for boot sequence, running preflight checks.","time":"2017-05-25T13:34:39.493Z","v":0}
{"name":"wings","hostname":"9320dd4fbd6e","pid":17,"server":"67d2f846-b348-4ece-bb6c-2b0e7ff2d65c","level":20,"msg":"No configuration object was assigned to this service. Skipping preflight.","time":"2017-05-25T13:34:39.493Z","v":0}
{"name":"wings","hostname":"9320dd4fbd6e","pid":17,"level":60,"err":{"message":"Cannot read property 'startup' of undefined","name":"TypeError","stack":"TypeError: Cannot read property 'startup' of undefined\n    at Async.parallel (/srv/daemon/src/services/index.js:195:83)\n    at /srv/daemon/node_modules/async/dist/async.js:3830:24\n    at eachOfArrayLike (/srv/daemon/node_modules/async/dist/async.js:1003:9)\n    at eachOf (/srv/daemon/node_modules/async/dist/async.js:1051:5)\n    at _parallel (/srv/daemon/node_modules/async/dist/async.js:3829:5)\n    at Object.parallelLimit [as parallel] (/srv/daemon/node_modules/async/dist/async.js:3912:5)\n    at Service.onConsole (/srv/daemon/src/services/index.js:152:15)\n    at Server.output (/srv/daemon/src/controllers/server.js:358:22)\n    at Carrier.carry.data (/srv/daemon/src/controllers/docker.js:225:29)\n    at emitOne (events.js:96:13)\n    at Carrier.emit (events.js:188:7)"},"msg":"A fatal error occured during an operation.","time":"2017-05-25T13:34:44.777Z","v":0}`

This is the behavior for the same server when attempting to stop the server:

`{"name":"wings","hostname":"9320dd4fbd6e","pid":17,"server":"67d2f846-b348-4ece-bb6c-2b0e7ff2d65c","level":30,"msg":"Server status has been changed to STOPPING","time":"2017-05-25T13:39:26.612Z","v":0}
{"name":"wings","hostname":"9320dd4fbd6e","pid":17,"level":60,"err":{"message":"Cannot read property 'stop' of undefined","name":"TypeError","stack":"TypeError: Cannot read property 'stop' of undefined\n    at Server.stop (/srv/daemon/src/controllers/server.js:329:41)\n    at Socket.activeSocket.on.data (/srv/daemon/src/http/socket.js:83:37)\n    at emitOne (events.js:96:13)\n    at Socket.emit (events.js:188:7)\n    at /srv/daemon/node_modules/socket.io/lib/socket.js:503:12\n    at _combinedTickCallback (internal/process/next_tick.js:67:7)\n    at process._tickDomainCallback (internal/process/next_tick.js:122:9)"},"msg":"A fatal error occured during an operation.","time":"2017-05-25T13:39:26.612Z","v":0}`

This is what happens when I kill the server:

`{"name":"wings","hostname":"9320dd4fbd6e","pid":17,"server":"67d2f846-b348-4ece-bb6c-2b0e7ff2d65c","level":30,"msg":"Server status has been changed to OFF","time":"2017-05-25T14:05:34.714Z","v":0}
{"name":"wings","hostname":"9320dd4fbd6e","pid":17,"level":60,"err":{"message":"Cannot read property 'log' of undefined","name":"TypeError","stack":"TypeError: Cannot read property 'log' of undefined\n    at Socket.activeSocket.on (/srv/daemon/src/http/socket.js:63:66)\n    at emitNone (events.js:86:13)\n    at Socket.emit (events.js:185:7)\n    at /srv/daemon/node_modules/socket.io/lib/socket.js:503:12\n    at _combinedTickCallback (internal/process/next_tick.js:67:7)\n    at process._tickDomainCallback (internal/process/next_tick.js:122:9)"},"msg":"A fatal error occured during an operation.","time":"2017-05-25T14:05:38.551Z","v":0}`

Now let me blow your mind.

When I create a new Call of Duty 4 server from within the panel, it starts and stops perfectly fine. I can't say nothing has changed. Obviously I updated the daemon. From a panel perspective, I did make one change: I had been experiencing issues **starting** my containers due to a bug where it would get stuck in a "suspended" state. To resolve this, I was instructed to add "true" to my "Scripts" section. After installing the new daemon, I had removed this from the "Scripts" section. Personally, I'm not sure if this would affect any servers that were previously created with this value existing. Additionally, this only affected the **creation** of servers, which is currently fine (or appears to be as of this post).

Here is the successful behavior of starting/stopping a brand new server:

`{"name":"wings","hostname":"9320dd4fbd6e","pid":17,"server":"4f10eff5-becd-4c12-924e-f701c6bfbcdd","level":30,"msg":"Server status has been changed to STARTING","time":"2017-05-25T13:43:50.267Z","v":0}
{"name":"wings","hostname":"9320dd4fbd6e","pid":17,"server":"4f10eff5-becd-4c12-924e-f701c6bfbcdd","level":20,"msg":"Checking size of server folder before booting.","time":"2017-05-25T13:43:50.267Z","v":0}
{"name":"wings","hostname":"9320dd4fbd6e","pid":17,"server":"4f10eff5-becd-4c12-924e-f701c6bfbcdd","level":20,"msg":"Initializing for boot sequence, running preflight checks.","time":"2017-05-25T13:43:50.272Z","v":0}
{"name":"wings","hostname":"9320dd4fbd6e","pid":17,"server":"4f10eff5-becd-4c12-924e-f701c6bfbcdd","level":30,"msg":"Server status has been changed to ON","time":"2017-05-25T13:43:57.404Z","v":0}
{"name":"wings","hostname":"9320dd4fbd6e","pid":17,"server":"4f10eff5-becd-4c12-924e-f701c6bfbcdd","level":30,"msg":"Server status has been changed to STOPPING","time":"2017-05-25T13:44:35.457Z","v":0}
{"name":"wings","hostname":"9320dd4fbd6e","pid":17,"server":"4f10eff5-becd-4c12-924e-f701c6bfbcdd","level":30,"msg":"Server status has been changed to OFF","time":"2017-05-25T13:44:45.494Z","v":0}`

So tell me why, when the server has the same exact main.json and the same exact "done" and "stop" configuration via the panel, why this one works and the others do not? What is breaking there?

Finally, I can no longer create Minecraft servers. They error out upon creation:

`{"name":"wings","hostname":"9320dd4fbd6e","pid":17,"server":"fca7d70e-6973-4729-9dd6-93ce23db213a","level":20,"msg":"Pulling alpine:3.4 image if it is not already on the system.","time":"2017-05-25T14:17:05.068Z","v":0}
{"name":"wings","hostname":"9320dd4fbd6e","pid":17,"level":60,"err":{"message":"callback is not a function","name":"TypeError","stack":"TypeError: callback is not a function\n    at Async.auto.close_stream.callback (/srv/daemon/src/controllers/option.js:101:24)\n    at runTask (/srv/daemon/node_modules/async/dist/async.js:1619:13)\n    at /srv/daemon/node_modules/async/dist/async.js:1559:13\n    at processQueue (/srv/daemon/node_modules/async/dist/async.js:1569:13)\n    at taskComplete (/srv/daemon/node_modules/async/dist/async.js:1588:9)\n    at /srv/daemon/node_modules/async/dist/async.js:1612:17\n    at /srv/daemon/node_modules/async/dist/async.js:906:16\n    at /srv/daemon/node_modules/graceful-fs/graceful-fs.js:99:16\n    at /srv/daemon/node_modules/graceful-fs/graceful-fs.js:43:10\n    at FSReqWrap.oncomplete (fs.js:123:15)"},"msg":"A fatal error occured during an operation.","time":"2017-05-25T14:17:05.069Z","v":0}
{"name":"wings","hostname":"9320dd4fbd6e","pid":17,"level":30,"msg":"Pulling image alpine:3.4 ... this could take a few minutes.","time":"2017-05-25T14:17:20.752Z","v":0}
{"name":"wings","hostname":"9320dd4fbd6e","pid":17,"server":"fca7d70e-6973-4729-9dd6-93ce23db213a","level":30,"msg":"Temporarily suspending server to avoid mishaps...","time":"2017-05-25T14:17:20.753Z","v":0}
{"name":"wings","hostname":"9320dd4fbd6e","pid":17,"server":"fca7d70e-6973-4729-9dd6-93ce23db213a","level":20,"msg":"Server is has been marked as requiring a rebuild on next boot cycle.","time":"2017-05-25T14:17:20.754Z","v":0}
{"name":"wings","hostname":"9320dd4fbd6e","pid":17,"server":"fca7d70e-6973-4729-9dd6-93ce23db213a","level":40,"msg":"Server has been suspended.","time":"2017-05-25T14:17:20.792Z","v":0}`

Here's a summary that I will update as things change:

Was working in 0.4.1 develop (2017/05/25 10:35 AM EST):

- Can create Call of Duty 4 servers
- Can create Call of Duty 2 servers
- Call of Duty 2 servers created after daemon start/stop fine
- Call of Duty 4 servers created after daemon start/stop fine
- Minecraft servers created before/after daemon start/stop fine

Broken in 0.4.1 develop (2017/05/25 10:35 AM EST):

- Minecraft servers cannot be created (stuck in "suspended" state downloading Alpine image)
~- Call of Duty 4 servers created prior to updating the daemon can no longer be stopped/started via the panel~ Note: As mentioned below, I removed all of these therefore preventing us from testing.

Resolved in 0.4.1 official (2017/05/25 11:50 AM EST):

- Minecraft servers are able to be created successfully

From my perspective, there's no real rhyme or reason to this but I'd like to keep this issue open at least as a line of communication for error reporting on this particular issue. I am going to try to document and log any and every change I make to see if we can get to the bottom of this. Am I using a daemon in a Docker container? Yes. Do I think this could be contributing to some of the trouble? Perhaps. I like to consider myself a heavier user of this panel than most though, so I expect regardless of how I do this I will still be running into little bugs here and there. Thus, I'm going to try to document as much as possible using my existing environment. If necessary, I may spin up a secondary test environment without the use of containers.

**Comment by xyeLz** [comment]
Changelog:

- I removed all Call of Duty 4 servers that were created during the lifetime of previous version of the daemon (these are the ones that couldn't start). I figure it would make more sense to work with current servers and the current version rather than dealing with the past version and trying to reverse engineer why it may be broken. Therefore, we can no longer troubleshoot the Call of Duty 4 servers created prior to updating the daemon. However, any future errors that occur I will be documenting.
- I removed my existing wings container and image and have downloaded the latest daemon container using the one created by @parkervcp which utilizes node. After doing this, I am able to create a Minecraft server again without any issues, resolving my second issue stated above.
